File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/settings/display/DisplaySettingsPresenter.java
package com.github.anrimian.musicplayer.ui.settings.display;

import com.github.anrimian.musicplayer.domain.business.settings.DisplaySettingsInteractor;

import io.reactivex.Observable;
import io.reactivex.disposables.CompositeDisposable;
import moxy.InjectViewState;
import moxy.MvpPresenter;

@InjectViewState
public class DisplaySettingsPresenter extends MvpPresenter<DisplaySettingsView> {

    private final DisplaySettingsInteractor interactor;

    private final CompositeDisposable presenterDisposable = new CompositeDisposable();

    public DisplaySettingsPresenter(DisplaySettingsInteractor interactor) {
        this.interactor = interactor;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        getViewState().showCoversChecked(interactor.isCoversEnabled());
        getViewState().showCoversInNotificationChecked(interactor.isCoversInNotificationEnabled());
        getViewState().showColoredNotificationChecked(interactor.isColoredNotificationEnabled());
        getViewState().showCoversOnLockScreenChecked(interactor.isCoversOnLockScreenEnabled());

        subscribeOnCoversEnabledState();
        subscribeOnColoredNotificationEnabledState();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenterDisposable.dispose();
    }

    void onCoversChecked(boolean checked) {
        getViewState().showCoversChecked(checked);
        interactor.setCoversEnabled(checked);
    }

    void onCoversInNotificationChecked(boolean checked) {
        getViewState().showCoversInNotificationChecked(checked);
        interactor.setCoversInNotificationEnabled(checked);
    }

    void onColoredNotificationChecked(boolean checked) {
        getViewState().showColoredNotificationChecked(checked);
        interactor.setColoredNotificationEnabled(checked);
    }

    void onCoversOnLockScreenChecked(boolean checked) {
        getViewState().showCoversOnLockScreenChecked(checked);
        interactor.setCoversOnLockScreenEnabled(checked);
    }

    private void subscribeOnColoredNotificationEnabledState() {
        presenterDisposable.add(Observable.combineLatest(interactor.getCoversEnabledObservable(),
                interactor.getCoversInNotificationEnabledObservable(),
                (covers, notification) -> covers && notification)
                .subscribe(getViewState()::showColoredNotificationEnabled));
    }

    private void subscribeOnCoversEnabledState() {
        presenterDisposable.add(interactor.getCoversEnabledObservable()
                .subscribe(this::onCoversEnabled));
    }

    private void onCoversEnabled(boolean enabled) {
        getViewState().showCoversInNotificationEnabled(enabled);
        getViewState().showShowCoversOnLockScreenEnabled(enabled);
    }
}

File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/settings/display/DisplaySettingsPresenter.kt
package com.github.anrimian.musicplayer.ui.settings.display

import com.github.anrimian.musicplayer.domain.interactors.settings.DisplaySettingsInteractor
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser
import com.github.anrimian.musicplayer.ui.common.mvp.AppPresenter
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Scheduler

class DisplaySettingsPresenter(private val interactor: DisplaySettingsInteractor,
                               uiScheduler: Scheduler,
                               errorParser: ErrorParser
) : AppPresenter<DisplaySettingsView>(uiScheduler, errorParser) {

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        viewState.showCoversChecked(interactor.isCoversEnabled)
        viewState.showCoversInNotificationChecked(interactor.isCoversInNotificationEnabled)
        viewState.showColoredNotificationChecked(interactor.isColoredNotificationEnabled)
        viewState.showCoversOnLockScreenChecked(interactor.isCoversOnLockScreenEnabled)
        subscribeOnCoversEnabledState()
        subscribeOnColoredNotificationEnabledState()
    }

    fun onCoversChecked(checked: Boolean) {
        viewState.showCoversChecked(checked)
        interactor.isCoversEnabled = checked
    }

    fun onCoversInNotificationChecked(checked: Boolean) {
        viewState.showCoversInNotificationChecked(checked)
        interactor.isCoversInNotificationEnabled = checked
    }

    fun onColoredNotificationChecked(checked: Boolean) {
        viewState.showColoredNotificationChecked(checked)
        interactor.isColoredNotificationEnabled = checked
    }

    fun onCoversOnLockScreenChecked(checked: Boolean) {
        viewState.showCoversOnLockScreenChecked(checked)
        interactor.isCoversOnLockScreenEnabled = checked
    }

    private fun subscribeOnColoredNotificationEnabledState() {
        Observable.combineLatest(interactor.coversEnabledObservable,
                interactor.coversInNotificationEnabledObservable,
                { covers, notification -> covers && notification })
                .unsafeSubscribeOnUi(viewState::showColoredNotificationEnabled)
    }

    private fun subscribeOnCoversEnabledState() {
        interactor.coversEnabledObservable.unsafeSubscribeOnUi(this::onCoversEnabled)
    }

    private fun onCoversEnabled(enabled: Boolean) {
        viewState.showCoversInNotificationEnabled(enabled)
        viewState.showShowCoversOnLockScreenEnabled(enabled)
    }
}